pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,313 @@
+	}
+	if err := p.updateSpotPricing(ctx); err != nil {
+		logging.FromContext(ctx).Errorf("updating spot pricing, %s, using existing pricing data from %s", err, p.spotUpdateTime.Format(time.RFC822Z))
+	}
+}
+
+func (p *PricingProvider) updateOnDemandPricing(ctx context.Context) error {
+	prices := map[string]float64{}
+	// standard on-demand instances
+	err := p.fetchOnDemandPricing(ctx, prices,
+		&pricing.Filter{
+			Field: aws.String("tenancy"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Shared"),
+		},
+		&pricing.Filter{
+			Field: aws.String("productFamily"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Compute Instance"),
+		})
+	if err != nil {
+		return err
+	}
+
+	// bare metal on-demand prices
+	err = p.fetchOnDemandPricing(ctx, prices,
+		&pricing.Filter{
+			Field: aws.String("tenancy"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Dedicated"),
+		},
+		&pricing.Filter{
+			Field: aws.String("productFamily"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Compute Instance (bare metal)"),
+		})
+	if err != nil {
+		return err
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.onDemandPrices = prices
+	p.onDemandUpdateTime = time.Now()
+	logging.FromContext(ctx).Infof("updated on-demand pricing with %d instance types", len(p.onDemandPrices))
+	return nil
+}
+
+func (p *PricingProvider) fetchOnDemandPricing(ctx context.Context, prices map[string]float64, additionalFilters ...*pricing.Filter) error {
+	filters := append([]*pricing.Filter{
+		{
+			Field: aws.String("regionCode"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String(p.region),
+		},
+		{
+			Field: aws.String("serviceCode"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("AmazonEC2"),
+		},
+		{
+			Field: aws.String("preInstalledSw"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("NA"),
+		},
+		{
+			Field: aws.String("operatingSystem"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Linux"),
+		},
+		{
+			Field: aws.String("capacitystatus"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Used"),
+		},
+		{
+			Field: aws.String("marketoption"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("OnDemand"),
+		}},
+		additionalFilters...)
+	return p.pricing.GetProductsPagesWithContext(ctx, &pricing.GetProductsInput{
+		Filters:     filters,
+		ServiceCode: aws.String("AmazonEC2")}, p.onDemandPage(prices))
+}
+
+// turning off cyclo here, it measures as a 12 due to all of the type checks of the pricing data which returns a deeply
+// nested map[string]interface{}
+// nolint: gocyclo
+func (p *PricingProvider) onDemandPage(prices map[string]float64) func(output *pricing.GetProductsOutput, b bool) bool {
+	return func(output *pricing.GetProductsOutput, b bool) bool {
+		for _, outer := range output.PriceList {
+			product, ok := outer["product"].(map[string]interface{})
```

Comment:
I felt bad writing this code.

Reply:
do you think it would be better to convert to json and then unmarshal to a struct?  You couldn't clean-up all of it with that, but might make the structure easy to see as the reader and limit all the `!ok` checks

Reply:
So we'd go []byte -> map[string]interface{} -> []byte -> new custom struct 😆   I could go either way, as long as the pricing API doesn't change much I think this is ok.  I suspect the API response is basically set in stone or everyone would be breaking.

Reply:
+1 to structifying this.